software/Test_Robot_Software/main.py

```diff
@@ -0,0 +1,93 @@
+from camera import RealsenseCamera
+from image_processor import ImageProcessor
+from api import RobotMovement
+from robot_state import State
+from state_machine import StateMachine
+from time import time
+
+import cv2
+
+class Main:
+    """
+    This class combines all tasks:
+    finding a ball, following a ball, etc
+    """
+
+    def __init__(self, robot_movement):
+
+        self.debug = True
+        self.cam = RealsenseCamera(exposure=100)
+        self.processor = ImageProcessor(self.cam, debug=self.debug)
+        self.processor.start()        
+
+        self.robot_movement = RobotMovement()
+        self.state_machine = StateMachine(self.robot_movement)
+        self.current_state = State.FIND_A_BALL
+        self.run = True
+
+if __name__ == "__main__":
+    main = Main(robot_movement=RobotMovement())
+
+    start = time()
+    fps = 0
+    frame = 0
+    frame_cnt = 0
+
+    try:
+        while True:
+            processed_data = main.processor.process_frame(aligned_depth=False)           
+            
+            if processed_data.balls:
```

Comment:
I would suggest to rewrite the state machine to take the processed inputs as an argument so you can call:

`state_machine.run_current_state(processed_data)`

Then the state machine can check if there are any balls and whatnot internally and change its state accordingly. Then this main basically needs to run the loop and provide data to it. and maybe draw some debug stuff

Reply:
updated the code accordingly